src/linalg.jl

```diff
@@ -629,6 +629,60 @@ function dot(A::AbstractSparseMatrixCSC, B::Union{DenseMatrixUnion,WrapperMatrix
     return conj(dot(B, A))
 end
 
+function dot(x::AbstractSparseVector, D::Diagonal, y::AbstractVector)
```

Comment:
Do you want to add a specialization for when `y` is sparse? Like

```julia
dot(x::AbstractVector, D::Diagonal, y::AbstractSparseVector) = adjoint(dot(y, D', x))
```

? Though this may introduce an infinite loop when `D'` returns an `Adjoint`. Or write it out?

Reply:
> Though this may introduce an infinite loop when D' returns an Adjoint. Or write it out?

It shouldn't no, D' returns a diagonal without wrapping an Adjoint type on top

Reply:
But for arrays of arrays it wraps an `Adjoint` around, and 3-arg `dot` with a middle `Adjoint` unwraps the adjoint again. I'm never sure anyone uses array of arrays "in production", but technically one could, and `dot` highlights that it acts recursively in such situations. It's just a few lines of code, so I think it's safer to spell it out.

Reply:
> But for arrays of arrays it wraps an Adjoint around, and 3-arg dot with a middle Adjoint unwraps the adjoint again

I don't get the issue, if `D'` is always guaranteed not to return an adjoint?

`adjoint(D::Diagonal) = Diagonal(adjoint.(D.diag))` is the only potentially implementation for recursive arrays, and it still returns a diagonal, it could potentially wrap around an adjoint diagonal but this doesn't trigger a recursive loop?

Reply:
You're right, sorry, I misread it (or misremembered).